File: desktop/src/lib/permissions.ts
import type { User } from "@/types";

export type PermissionAction = "can_view" | "can_create" | "can_edit" | "can_delete" | "can_export";

export function hasPermission(
  user: User | null,
  module: string,
  action: PermissionAction,
): boolean {
  if (!user) return false;
  if (user.role_code === "SUPER_ADMIN") return true;
  if (!user.permissions) return false;
  const perm = user.permissions.find((p) => p.module === module);
  if (!perm) return false;
  return perm[action] === 1;
}

export function hasAnyPermission(
  user: User | null,
  module: string,
  actions: PermissionAction[],
): boolean {
  return actions.some((action) => hasPermission(user, module, action));
}

File: android/app/src/main/java/com/gsoft/opus/core/Permissions.kt
package com.gsoft.opus.core

import com.gsoft.opus.domain.model.User

enum class PermissionAction { VIEW, CREATE, EDIT, DELETE, EXPORT }

/**
 * Mirrors the desktop `hasPermission` helper: SUPER_ADMIN bypasses the
 * permission table, otherwise the role permission for the given module
 * must explicitly allow the action.
 */
fun hasPermission(user: User?, module: String, action: PermissionAction): Boolean {
    if (user == null) return false
    if (user.roleCode == "SUPER_ADMIN") return true
    val perm = user.permissions.find { it.module == module } ?: return false
    return when (action) {
        PermissionAction.VIEW -> perm.canView
        PermissionAction.CREATE -> perm.canCreate
        PermissionAction.EDIT -> perm.canEdit
        PermissionAction.DELETE -> perm.canDelete
        PermissionAction.EXPORT -> perm.canExport
    }
}
